Handle 'COPY --from' in Dockerfile

TomSweeneyRedHat opened the pull request:
Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

Handle the "COPY --from" directives in a Dockerfile.  If we discover this line in our Dockerfiles, the code will create a  new FROM directive for the listed image which will cause the image to be pulled prior to the COPY command is executed.  Prior this was not happening.

Fixes: #1175 

Testing:
```
# cat ~/Dockerfile.copyfrom
FROM php:7.2 
COPY --from=composer:latest /usr/bin/composer /usr/local/bin/composer

# buildah bud -f ~/Dockerfile.copyfrom -t tom .
STEP 1: FROM composer:latest
Getting image source signatures
Copying blob sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde
 2.10 MiB / 2.10 MiB [======================================================] 0s
Copying blob sha256:5cf60c0c0763e6e5b51f7b06d1a681fbc422c4feadba6e28e0354de32fe18743
 1.29 MiB / 1.29 MiB [======================================================] 0s
Copying blob sha256:468219aa46893af0196f1629ab668d8120e76afb69af0c3cc51a0c0caeff594e
 1.22 KiB / 1.22 KiB [======================================================] 0s
Copying blob sha256:4a8b1a3b9e510caf57ea183cacd59ad9f85a026dac10eb86f68fec44d0dfda14
 169 B / 169 B [============================================================] 0s
Copying blob sha256:da2c88c4ca27c759936b47b3d60f6bc7e1996016686abb20992caa9d23e63ad1
 11.60 MiB / 11.60 MiB [====================================================] 1s
Copying blob sha256:c4f408a591350b599d1d9e2f510961e37188d566263750a6fefd764410ec2406
 497 B / 497 B [============================================================] 0s
Copying blob sha256:d65ea60b5d6c1c72c736fe5c73e87c2546ea7a7ff3016aa9742096d4d2b7ac83
 15.27 MiB / 15.27 MiB [====================================================] 2s
Copying blob sha256:7e266b5abba2ce2cf6c6db0e5db503e1929585ecafe556360ba1d339776b890a
 2.13 KiB / 2.13 KiB [======================================================] 0s
Copying blob sha256:bb41dfb37ff5079cb2e644061f6d0eead3bb3c34c9746962ae0624f7064e6718
 70.21 KiB / 70.21 KiB [====================================================] 0s
Copying blob sha256:e5082384099298f9946f5782b8e603127ee067a3dd5224e604a3fec001e16c2f
 25.74 MiB / 25.74 MiB [====================================================] 3s
Copying blob sha256:7b5f988bed2b85be5298c642cac66eb5e32d6d53ec987c95b94c14d27b8ba9da
 281 B / 281 B [============================================================] 0s
Copying blob sha256:f8ca7e00f959f6e0013eee5f8ece47dcf089f02edd77568a9e9e64850f9b4cdb
 1.44 MiB / 1.44 MiB [======================================================] 0s
Copying blob sha256:5820e42061a10c35d7caf3da604b1bcbb546d3686846595d2dee7cef1cce81c8
 464.26 KiB / 464.26 KiB [==================================================] 0s
Copying blob sha256:1da10be534f4a998d63ca161c9a56c9802c5a5b9bd105bef164132dff2c0f143
 534 B / 534 B [============================================================] 0s
Copying blob sha256:2c20a35aee8974e7135fdd1e0d6742a1b1a44498ce5c1a54770cceb99acbfd43
 92 B / 92 B [==============================================================] 0s
Copying config sha256:3aec6988164bfa2b2f4da94d80617a4727e05ec4a5cf71fd775bdc86fb9044df
 11.81 KiB / 11.81 KiB [====================================================] 0s
Writing manifest to image destination
Storing signatures
STEP 2: FROM php:7.2
Getting image source signatures
Copying blob sha256:a5a6f2f73cd8abbdc55d0df0d8834f7262713e87d6c8800ea3851f103025e0f0
 21.44 MiB / 21.44 MiB [====================================================] 3s
Copying blob sha256:633e0d1cd2a336ecc3124ec0b06e1039049f1716554038b9dfc8c3529f6fd988
 226 B / 226 B [============================================================] 0s
Copying blob sha256:fcdfdf7118ba156cb7115b991210623e82c6ad0f57aa55e38790f2bd734c50fa
 64.30 MiB / 64.30 MiB [====================================================] 9s
Copying blob sha256:4e7dc76b176992c0aa2fd3f3bc438ac6be98897e6025b68e8652b24b132ec952
 183 B / 183 B [============================================================] 0s
Copying blob sha256:35f4eff7ab0f9f7d1e4c6cc974f7e37f69249ce4ffb3e1d07347d387ba8e30a7
 11.89 MiB / 11.89 MiB [====================================================] 1s
Copying blob sha256:1f1e528293cb2b479a302698e8953c9c1326d512429229a3561d3b5a0372780b
 500 B / 500 B [============================================================] 0s
Copying blob sha256:d84e6ef5c1c71eb39c9541689cbf9bf1e01d8f2266e12ccf0988418a7d9f6166
 27.32 MiB / 27.32 MiB [====================================================] 4s
Copying blob sha256:e5a84d39467f707f84e67022d92848f8a2f32ed9881404cc4e7fc8ee3240a741
 2.14 KiB / 2.14 KiB [======================================================] 0s
Copying blob sha256:3557589ef6f988caed2046bd32a716193a523a371cb4d9b561730057660e0a1c
 259 B / 259 B [============================================================] 0s
Copying config sha256:8473cbe51b2266629239b74891d0d76ca5c05e2fd82c2bd99a6152c14d46f2b8
 9.56 KiB / 9.56 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
STEP 3: COPY --from=composer:latest /usr/bin/composer /usr/local/bin/composer
STEP 4: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.mountopt=nodev,overlay.override_kernel_check=true]localhost/tom:latest
Getting image source signatures
Skipping fetch of repeat blob sha256:ef68f6734aa485edf13a8509fe60e4272428deaf63f446a441b79d47fc5d17d3
Skipping fetch of repeat blob sha256:4572a80a7a5e1f51c2946b7e16d3a3e53a904686443fb900b104a166462bb4c0
Skipping fetch of repeat blob sha256:abae6a338e5cda66bcf4504c3f12b827947feb0b90c5c07d451eab49414e9b79
Skipping fetch of repeat blob sha256:96db4ce698adeefc1f4323b5bf05f5757c000fc5994bf09be6be6421adf9d6b6
Skipping fetch of repeat blob sha256:023f8b788c2ae2aa74b8ab21ac64023c1f29817a2e2e9c02f1d06800602dfd1d
Skipping fetch of repeat blob sha256:8ac80e6aa5a8f53ca1dd9e23f9739c5111987af9e2286939f4ca283db1d6d757
Skipping fetch of repeat blob sha256:4bfcbdab35067d10eb734598fb76ef1942a131b6ab5fcbcc50a3be848e66cd09
Skipping fetch of repeat blob sha256:f8f930b183b29346398273a3dea7b9bfa44ec1788c017bf0bd319fe64c71428d
Skipping fetch of repeat blob sha256:666ad3febf9c4bba841fddc50a59f351df6380b156f7c07294040f12ae4b146e
Copying blob sha256:93acf3a5effc72ca8e8ef46fab6e38c5be853c91e5ddc6cbe31f6411b7adcfb2
 464.30 KiB / 464.30 KiB [==================================================] 0s
Copying config sha256:75fb7d9d3699daf2eac77bf79ab041d038de47cd3715607c219215074f7616de
 8.13 KiB / 8.13 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
--> 75fb7d9d3699daf2eac77bf79ab041d038de47cd3715607c219215074f7616de
[root@localhost tests]# /usr/local/go/bin/go version
go version go1.10 linux/amd64
```

TomSweeneyRedHat commented:
bot, retest this please

TomSweeneyRedHat commented:
Going to need to step back on this one.  
I'm running into issues handling a Dockerfile like:
```
FROM scratch AS myname
COPY test.txt /

FROM scratch
COPY --from=myname /test.txt /etc/test.txt
```
as this change didn't consider that FROM has already been completed for the `COPY --from` construct.  Both instances should work.  With this change I first add a bare 'FROM myname` to the top of the Dockerfile and that fails.

rhatdan commented:
@TomSweeneyRedHat Needs a rebase.  Are you still working on this?

TomSweeneyRedHat commented:
Yes, still actively working.  This is one of those dike-plugging boys PR's, I get it working for one type of Dockerfile and then another breaks in the tests.  Mad mess of dependencies between Buildah and imagebuilder.

rhatdan commented:
Well @TomSweeneyRedHat now it needs a rebase.

rhatdan commented:
LGTM
@vrothberg @giuseppe @umohnani8 @nalind @pixdrift PTAL

giuseppe commented on imagebuildah/build.go:
```diff
@@ -1415,6 +1427,90 @@ func BuildDockerfiles(ctx context.Context, store storage.Store, options BuildOpt
+	for _, dfile := range dockerfiles {
+		if dfileBinary, err := ioutil.ReadAll(dfile); err == nil {
+			dfileString := fmt.Sprintf("%s", dfileBinary)
+			copyFromContent := strings.Split(dfileString, "COPY --from=")
```
should we use a regex here? It seems to not handle Dockerfiles when tabs are used (e.g. `COPY \t --from=myname\n`). The same Dockerfile works with Docker

TomSweeneyRedHat replied:
Yet another twist, thought I'd found them all.  Great catch @giuseppe, will see what I can do with it.  I was hoping to avoid a regex, but that might be the answer.

vrothberg replied:
I think a good regex for that is `"^\s*COPY\s+--from=([\S]+)\s*$"`

vrothberg commented:
@TomSweeneyRedHat, a meta comment on the commit/PR message. `Addresses` won't close the referenced issue automatically when this PR gets merged. You can pick one of the keywords mentioned here: https://help.github.com/articles/closing-issues-using-keywords/

Personally, I prefer `Fixes: ...` as it aligns (mostly) with the Kernel guidelines.

rhatdan commented:
@TomSweeneyRedHat Still working on this?

TomSweeneyRedHat commented:
@rhatdan yep, set it aside yesterday to look at the memory swap issue and other stuff.  Just jumped back in a bit ago trying to get regexp's to work.  I miss Perl atm.

TomSweeneyRedHat commented:
Per @giuseppe and @vrothberg 's comments, I've converted to use a reexp to parse the strings.  The bonus is that enables me to condense the two loops for ' as|AS ' into one loop.  To make sure we have a test for both instances, I copied the multi-stage-builds test suite to multi-stage-builds-small-as and kept the ' as ' small there and changed ' as ' to ' AS ' in the original.

Tests passed in my sandbox, hopefully the same in the CI system.

vrothberg commented on imagebuildah/build.go:
```diff
@@ -1421,6 +1434,78 @@ func BuildDockerfiles(ctx context.Context, store storage.Store, options BuildOpt
+	for _, dfile := range dockerfiles {
+		if dfileBinary, err := ioutil.ReadAll(dfile); err == nil {
+			dfileString := fmt.Sprintf("%s", dfileBinary)
+			copyRE := regexp.MustCompile(`\s*COPY\s+--from=`)
```
Can we move the regexes outside this function to avoid them being recompiled for each `dfile`? I didn't test it (about to take the train to the Munich office) but could trailing whitespaces/tabs/etc. cause problems here? If so, I suggest using a group for matching instead of splitting.

TomSweeneyRedHat commented:
Per @giuseppe and @vrothberg 's comments, I've converted to use a reexp to parse the strings.  The bonus is that enables me to condense the two loops for ' as|AS ' into one loop.  To make sure we have a test for both instances, I copied the multi-stage-builds test suite to multi-stage-builds-small-as and kept the ' as ' small there and changed ' as ' to ' AS ' in the original.

Tests passed in my sandbox, hopefully the same in the CI system.

TomSweeneyRedHat commented:
@vrothberg I've updated and moved the compiles outside of the for loop, good thought.  I know enough about groups to be dangerous and didn't see a huge buy with them.  However, I did add some TrimSpace() calls to handle something like "COPY --from={tab}myname".  With that in play we now fail in a similar manner as Docker.  

Docker:
```
Step 6/7 : COPY --from=	myname /Dockerfile.name /Dockerfile.name
COPY failed: stat /var/lib/docker/tmp/docker-builder707777720/myname: no such file or directory
```

Buildah
```
STEP 8: COPY --from=	myname /Dockerfile.name /Dockerfile.name
error building at step {Env:[ PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:copy Args:[myname /Dockerfile.name /Dockerfile.name] Flags:[--from=] Attrs:map[] Message:COPY --from= myname /Dockerfile.name /Dockerfile.name Original:COPY --from=	myname /Dockerfile.name /Dockerfile.name}: destination "/var/lib/containers/storage/overlay/70c29005c032cdce2c216a3646cd7303305c97c64d8acbcbccc9d2d0a87a4b8a/merged/Dockerfile.name" is not a directory
```

Unless you've huge concerns, as this PR fixes the mainline case, I'd like to get this in (assuming happy tests and no mess ups on the new code) and then we can pick at the outliers further if you thinks it's useful/more efficient.

vrothberg commented:
LGTM, thanks a lot @TomSweeneyRedHat

vrothberg commented:
@rh-atomic-bot r+
